### PR TITLE
Convert Eio.Process to FCMs

### PR DESCRIPTION
**Pull request by patricoferris**
Started working with processes and thought I'd convert them to FCMs whilst I'm here :))

### CODE REVIEW

File: lib_eio/unix/eio_unix.ml [hunk 1]
```diff
@@ -39,7 +39,7 @@ module Stdenv = struct
     stderr : sink_ty r;
     net : [`Unix | `Generic] Eio.Net.ty r;
     domain_mgr : Eio.Domain_manager.t;
-    process_mgr : Process.mgr;
+    process_mgr : Process.mgr_ty Process.mgr;
```
Comment (talex5) [inline lib_eio/unix/eio_unix.ml:42]: For consistency with net above, and a bit simpler:
```suggestion
    process_mgr : Process.mgr_ty r;
```

File: lib_eio/process.ml [hunk 2]
```diff
@@ -48,21 +48,36 @@ let await_exn ?(is_success = Int.equal 0) proc =
...
-    sw:Switch.t ->
-    ?cwd:Fs.dir_ty Path.t ->
-    ?stdin:Flow.source_ty r ->
-    ?stdout:Flow.sink_ty r ->
-    ?stderr:Flow.sink_ty r ->
-    ?env:string array ->
-    ?executable:string ->
-    string list ->
-    t
+type mgr_ty = [`Mgr]
```
Comment (talex5) [inline lib_eio/process.ml:51]: These tags have to be globally unique:
```suggestion
type mgr_ty = [`Process_mgr]
```

File: lib_eio/process.ml [hunk 3]
```diff
@@ -48,21 +48,36 @@ let await_exn ?(is_success = Int.equal 0) proc =
...
+type 'a mgr = ([> mgr_ty] as 'a) r
+
+module Pi = struct
+  module type MGR = sig
+    val pipe : 
+      sw:Switch.t ->
```
Comment (talex5) [inline lib_eio/process.ml:57]:
```suggestion
    val pipe :
      t ->
      sw:Switch.t ->
```

File: lib_eio/process.ml [hunk 4]
```diff
@@ -48,21 +48,36 @@ let await_exn ?(is_success = Int.equal 0) proc =
...
+      [Flow.source_ty | Resource.close_ty] r * [Flow.sink_ty | Resource.close_ty] r
+
+    val spawn :
+      sw:Switch.t ->
```
Comment (talex5) [inline lib_eio/process.ml:61]:
```suggestion
    val spawn :
      t ->
      sw:Switch.t ->
```

File: lib_eio/process.ml [hunk 5]
```diff
@@ -48,21 +48,36 @@ let await_exn ?(is_success = Int.equal 0) proc =
...
+      ?env:string array ->
+      ?executable:string ->
+      string list ->
+      t
+
+    type t
+  end
+
+  type (_, _, _) Resource.pi +=
+    | Proc : ('t, (module MGR with type t = 't), [> mgr_ty]) Resource.pi
```
Comment (talex5) [inline lib_eio/process.ml:75]:
```suggestion
    | Mgr : ('t, (module MGR with type t = 't), [> mgr_ty]) Resource.pi
```

File: lib_eio/unix/process.ml [hunk 6]
```diff
@@ -69,23 +69,49 @@ let get_env = function
   | Some e -> e
   | None -> Unix.environment ()
 
-class virtual mgr = object (self)
-  inherit Eio.Process.mgr
+type mgr_ty = [`Mgr_unix | Eio.Process.mgr_ty]
```
Comment (talex5) [inline lib_eio/unix/process.ml:72]: We might want to copy the ``[`Platform of 'tag | ...]`` pattern used in net. That allows saying that a Unix process manager spawns Unix processes.

**Comment by patricoferris** [comment]
Tried to follow the `Platform of tag` method in the latest commits... I might have got lost in the polymorphic variants...

**Review by talex5 — COMMENTED** [review]
I pushed some changes. In particular, I renamed the types a bit: `Process.t` is now a process. The process manager is now `Process.mgr`.

**Review by talex5 — APPROVED** [review]

**Pull request merged**